app/finders/text_search.rb

```diff
@@ -0,0 +1,36 @@
+class TextSearch
+  attr_reader :current_user, :current_account, :params
+
+  DEFAULT_STATUS = 'open'.freeze
+
+  def initialize(current_user, params)
+    @current_account = current_user.account || Current.account
+    @params = params
+  end
+
+  def perform
+    {
+      messages: filter_messages,
+      conversations: filter_conversations,
+      contacts: filter_contacts
+    }
+  end
+
+  private
+
+  def filter_conversations
+    conversation_ids = PgSearch.multisearch((@params[:q]).to_s).where(account_id: @current_account,
+                                                                      searchable_type: 'Conversation').pluck(:searchable_id)
+    @conversations = Conversation.where(id: conversation_ids)
```

Comment:
We should do pagination here. this could be the culprit behind the slow search for shorter terms. 
We can also look into doing a join query so that the pagination works effectively.  

The same goes for messages and contacts as well.

Reply:
`PgSearch.multisearch` joins with conversations? 

@sojan-official I have discussed this with @fayazara  pagination wouldn't work for a multipage search he said, instead, we are limiting the result to the top 20.

Reply:
@sojan-official MultiSearch generates query like this

```PgSearch.multisearch('test').to_sql
=> "SELECT \"pg_search_documents\".* FROM \"pg_search_documents\" INNER JOIN (SELECT \"pg_search_documents\".\"id\" AS pg_search_id, (ts_rank((to_tsvector('simple', coalesce(\"pg_search_documents\".\"content\"::text, ''))), (to_tsquery('simple', ''' ' || 'test' || ' ''')), 0)) AS rank FROM \"pg_search_documents\" WHERE ((to_tsvector('simple', coalesce(\"pg_search_documents\".\"content\"::text, ''))) @@ (to_tsquery('simple', ''' ' || 'test' || ' ''')))) AS pg_search_ce9b9dd18c5c0023f2116f ON \"pg_search_documents\".\"id\" = pg_search_ce9b9dd18c5c0023f2116f.pg_search_id ORDER BY pg_search_ce9b9dd18c5c0023f2116f.rank DESC, \"pg_search_documents\".\"id\" ASC"```

I will check if joins are possible.